### PR TITLE
Revamp modes

### PR DESCRIPTION
**Pull request by mbfraga**
## Summary / How this PR fixes the problem?
This PR moves more of the move logic to the actual TransformMode, cleaning up selectedboundmanager, canvas, etc. It also uses transformmode directly in iteminsertmode instead of relying on code duplication. I also cleaned up some code in the nob manager and made it a bit easier to get the right data. Still consider these things a wip, but it seems to be in a good state.

Also reworked the rotation code.

## Steps to Test
All transformations should work as expected. Snap guides should show and disappear properly.

## Known Issues / Things To Do
None that I can think of.

## This PR fixes/implements the following **bugs/features**:
This actually does help fix some rotation quirks.

### CODE REVIEW

File: src/Lib/Canvas.vala [hunk 1]
```diff
@@ -493,6 +496,20 @@ public class Akira.Lib.Canvas : Goo.Canvas {
...
+    private void on_update_snap_decorators () {
+        var extra_context = mode_manager.active_mode_extra_context ();
+        if (extra_context is Akira.Lib.Modes.TransformMode.TransformExtraContext) {
+            var tec = extra_context as Akira.Lib.Modes.TransformMode.TransformExtraContext;
+            snap_manager.generate_decorators (tec.snap_guide_data);
+        }
+        else if (snap_manager.is_active ()) {
+            snap_manager.reset_decorators ();
+        }
+    }
```
Comment (Alecaddd) [inline src/Lib/Canvas.vala:511]: nit code styling.

The else conditions is not really an alternative to the active extra context, so stylistically I prefer an early return and then a following condition. This gives a good sense of the priority in which we want to check for things.

File: src/Lib/Canvas.vala [hunk 2]
```diff
@@ -493,6 +496,20 @@ public class Akira.Lib.Canvas : Goo.Canvas {
...
     }
 
+    /*
+     * Will update snap decorators if necessary.
+     */
+    private void on_update_snap_decorators () {
+        var extra_context = mode_manager.active_mode_extra_context ();
+        if (extra_context is Akira.Lib.Modes.TransformMode.TransformExtraContext) {
+            var tec = extra_context as Akira.Lib.Modes.TransformMode.TransformExtraContext;
+            snap_manager.generate_decorators (tec.snap_guide_data);
```
Comment (Alecaddd) [inline src/Lib/Canvas.vala:506]: You could type cast this directly when generating the decorators:
`snap_manager.generate_decorators (((Lib.Modes.TransformMode.TransformExtraContext) extra_context).snap_guide_data);`

Also, in general, you can avoid using the `Akira.` namespace and make the code a little shorter.

File: src/Lib/Managers/ModeManager.vala [hunk 3]
```diff
@@ -127,6 +127,18 @@ public class Akira.Lib.Managers.ModeManager : Object {
...
 
+    /*
+     * Returns optional extra context. See InteractionMode for more details.
+     */
+    public Object? active_mode_extra_context () {
+        if (pan_mode != null) {
+            return pan_mode.extra_context ();
+        }
+
+        return (active_mode != null) ? active_mode.extra_context () : null;
```
Comment (Alecaddd) [inline src/Lib/Managers/ModeManager.vala:138]: I think the parenthesis are not required for a ternary condition.

File: src/Lib/Managers/ModeManager.vala [hunk 4]
```diff
@@ -127,6 +127,18 @@ public class Akira.Lib.Managers.ModeManager : Object {
...
+
```
Comment (Alecaddd) [inline src/Lib/Managers/ModeManager.vala:139]: nit: remove blank line.

File: src/Lib/Modes/InteractionMode.vala [hunk 5]
```diff
@@ -50,48 +50,58 @@ public interface Akira.Lib.Modes.InteractionMode : Object {
...
+    public virtual bool button_release_event (Gdk.EventButton event) { return false; }
 
     /*
      * Override to define mouse motion event. Return true to absorb.
      */
-    public abstract bool motion_notify_event (Gdk.EventMotion event);
+    public virtual bool motion_notify_event (Gdk.EventMotion event) { return false; }
+
+    /*
+     * Optionally override to provide some extra context that may be used.
```
Comment (Alecaddd) [inline src/Lib/Modes/InteractionMode.vala:98]: nit: remove period

File: src/Lib/Modes/InteractionMode.vala [hunk 6]
```diff
@@ -50,48 +50,58 @@ public interface Akira.Lib.Modes.InteractionMode : Object {
...
+     * further up to know how to update other managers based on this mode.
+     *
+     * This context should not be relied upon for vital opertions or anything
+     * that requires a specific order of operations. It is generally meant for
+     * more cosmetic featurs.
```
Comment (Alecaddd) [inline src/Lib/Modes/InteractionMode.vala:103]: nit: features

File: src/Lib/Modes/ItemInsertMode.vala [hunk 7]
```diff
@@ -39,62 +39,89 @@ public class Akira.Lib.Modes.ItemInsertMode : Object, InteractionMode {
...
     }
 
-    public bool key_release_event (Gdk.EventKey event) {
+    public override bool key_release_event (Gdk.EventKey event) {
+        if (transform_mode != null) {
+            return transform_mode.key_press_event (event);
+        }
         return false;
-    }
+     }
```
Comment (Alecaddd) [inline src/Lib/Modes/ItemInsertMode.vala:74]: nit: wrong indentation

File: src/Lib/Modes/ItemInsertMode.vala [hunk 8]
```diff
@@ -39,62 +39,89 @@ public class Akira.Lib.Modes.ItemInsertMode : Object, InteractionMode {
...
+        }
         return false;
     }
 
-    public bool key_release_event (Gdk.EventKey event) {
+    public override bool key_release_event (Gdk.EventKey event) {
+        if (transform_mode != null) {
+            return transform_mode.key_press_event (event);
+        }
         return false;
```
Comment (Alecaddd) [inline src/Lib/Modes/ItemInsertMode.vala:73]: maybe have this as an inline ternary operation?

File: src/Lib/Modes/TransformMode.vala [hunk 9]
```diff
@@ -24,57 +24,433 @@
...
+    public override void mode_begin () {
+        unowned var selected_items = canvas.selected_bound_manager.selected_items;
+        if (!initialize_items_drag_state (selected_items, ref initial_drag_state)) {
+            debug ("TransformMode only works if an item is selected");
+            if (mode_manager != null) {
+                mode_manager.deregister_mode (mode_type ());
+                return;
+            }
+        }
+    }
```
Comment (Alecaddd) [inline src/Lib/Modes/TransformMode.vala:84]: I'm not a fan of indented conditions.
Would be possible to revert the conditions and use early returns to keep 1 level of indentation?

File: src/Lib/Modes/TransformMode.vala [hunk 10]
```diff
@@ -24,57 +24,433 @@
...
         return true;
     }
 
-    public bool button_release_event (Gdk.EventButton event) {
-        mode_manager.deregister_mode (mode_type ());
+    public override bool button_release_event (Gdk.EventButton event) {
+        if (mode_manager != null) {
+            mode_manager.deregister_mode (mode_type ());
+        }
         return true;
```
Comment (Alecaddd) [inline src/Lib/Modes/TransformMode.vala:126]: ternary operation here as well.

File: src/Lib/Modes/TransformMode.vala [hunk 11]
```diff
@@ -24,57 +24,433 @@
...
+                    event.y,
+                    ref transform_extra_context.snap_guide_data
+                );
+                break;
+        }
+
+        // Notify the X & Y values in the state manager.
+        canvas.window.event_bus.reset_state_coords ();
+
+
```
Comment (Alecaddd) [inline src/Lib/Modes/TransformMode.vala:177]: nit: remove extra blank line.

File: src/Lib/Modes/TransformMode.vala [hunk 12]
```diff
@@ -24,57 +24,433 @@
...
+        drag_state.item_x = item.coordinates.x;
+        drag_state.item_y = item.coordinates.y;
+
+        item.get_transform (out drag_state.item_transform);
+
+        if (selected_items.length () == 1) {
+            drag_state.item_width = item.size.width;
+            drag_state.item_height = item.size.height;
+        } else {
+            // TODO there should probably be a nice method to get a bounding box from a list of items.
```
Comment (Alecaddd) [inline src/Lib/Modes/TransformMode.vala:207]: I should have this nailed down in the multiselect PR :smile:

File: src/Lib/Modes/TransformMode.vala [hunk 13]
```diff
@@ -24,57 +24,433 @@
...
+        // for now we only transform one item
+        Akira.Lib.Items.CanvasItem item = selected_items.nth_data (0);
+
+       bool snapping_allowed = true;
+       // Don't snap rotated items.
+       if (item.rotation != null && GLib.Math.fmod (item.rotation.rotation, 90) != 0) {
+           snapping_allowed = false;
+       }
+
+        int snap_offset_x = 0;
```
Comment (Alecaddd) [inline src/Lib/Modes/TransformMode.vala:328]: wrong indentation from here going down.

File: src/Lib/Modes/TransformMode.vala [hunk 14]
```diff
@@ -24,57 +24,433 @@
...
+            var step_num = GLib.Math.round (new_rotation / 15.0);
+            new_rotation = 15.0 * step_num;
+        }
+
+        Akira.Lib.Items.CanvasItem item = selected_items.nth_data (0);
+        // Cap new_rotation to the [0, 360] range.
+        new_rotation = GLib.Math.fmod (new_rotation + 360, 360);
+        item.rotation.rotation = Akira.Utils.AffineTransform.fix_size (new_rotation);
+    }
+
```
Comment (Alecaddd) [inline src/Lib/Modes/TransformMode.vala:455]: nit: remove blank line.

**Review by Alecaddd — CHANGES_REQUESTED** [review]
Since the PR is broken, I did a first quick swipe review from a pure code POV.
I like the improvements and the modularity. Great work!

File: src/Lib/Canvas.vala (hunk 1 above)
Comment (mbfraga) [reply src/Lib/Canvas.vala:511]: The tacit else is: Nothing wants snap guides, and there isn't an active one to reset; so don't do anything. I don't think an early return is the right fix here.

File: src/Lib/Managers/ModeManager.vala (hunk 3 above)
Comment (mbfraga) [reply src/Lib/Managers/ModeManager.vala:138]: It isn't but I think it's better for readability.But I changed it.

File: src/Lib/Canvas.vala (hunk 2 above)
Comment (mbfraga) [reply src/Lib/Canvas.vala:506]: I separate it more for readability (still made the change)

**Pull request merged**